extensions/data-plane/data-plane-aws-s3/src/main/java/org/eclipse/edc/connector/dataplane/aws/s3/S3DataSink.java

```diff
@@ -75,31 +82,46 @@ protected StreamResult<Object> transferParts(List<DataSource.Part> parts) {
 
                 client.completeMultipartUpload(CompleteMultipartUploadRequest.builder()
                         .bucket(bucketName)
-                        .key(keyName)
+                        .key(key)
                         .uploadId(uploadId)
                         .multipartUpload(CompletedMultipartUpload.builder()
                                 .parts(completedParts)
                                 .build())
                         .build());
 
             } catch (Exception e) {
-                return uploadFailure(e, keyName);
+                return uploadFailure(e, key);
             }
+            registerCompletedFile(key);
         }
 
         return StreamResult.success();
     }
 
+    String getDestinationObjectName(String partName) {
+        var name = !StringUtils.isNullOrEmpty(keyName) && StringUtils.isNullOrBlank(keyPrefix) ? keyName : partName;
+        if (!StringUtils.isNullOrEmpty(folderName)) {
+            name = folderName.endsWith("/") ? folderName + name : folderName + "/" + name;
+        }
+        return name;
+    }
```

Comment:
Verify if the `Consumer` chose to bundle objects (`Part`s) in a folder by checking the presence of `folderName` (or `keyPrefix` as you choose) to decide on the `partName` prefixing.